crates/optimism/trie/src/metrics.rs

```diff
@@ -0,0 +1,495 @@
+        &self.metrics
+    }
+}
+
+/// Wrapper for [`OpProofsTrieCursor`] that records metrics.
+#[derive(Debug)]
+pub struct TrieCursorWithMetrics<C> {
+    cursor: C,
+    metrics: Arc<StorageMetrics>,
+}
+
+impl<C> TrieCursorWithMetrics<C> {
+    /// Create a new cursor wrapper with metrics.
+    pub const fn new(cursor: C, metrics: Arc<StorageMetrics>) -> Self {
+        Self { cursor, metrics }
+    }
+}
+
+impl<C: OpProofsTrieCursor> OpProofsTrieCursor for TrieCursorWithMetrics<C> {
+    fn seek_exact(
+        &mut self,
+        path: Nibbles,
+    ) -> OpProofsStorageResult<Option<(Nibbles, BranchNodeCompact)>> {
+        self.metrics.record_operation(StorageOperation::TrieCursorSeekExact, || {
+            self.cursor.seek_exact(path)
+        })
+    }
+
+    fn seek(
+        &mut self,
+        path: Nibbles,
+    ) -> OpProofsStorageResult<Option<(Nibbles, BranchNodeCompact)>> {
+        self.metrics.record_operation(StorageOperation::TrieCursorSeek, || self.cursor.seek(path))
+    }
+
+    fn next(&mut self) -> OpProofsStorageResult<Option<(Nibbles, BranchNodeCompact)>> {
+        self.metrics.record_operation(StorageOperation::TrieCursorNext, || self.cursor.next())
+    }
+
+    fn current(&mut self) -> OpProofsStorageResult<Option<Nibbles>> {
+        self.metrics.record_operation(StorageOperation::TrieCursorCurrent, || self.cursor.current())
+    }
+}
+
+/// Wrapper for [`OpProofsHashedCursor`] that records metrics.
+#[derive(Debug)]
+pub struct HashedCursorWithMetrics<C> {
+    cursor: C,
+    metrics: Arc<StorageMetrics>,
+}
+
+impl<C> HashedCursorWithMetrics<C> {
+    /// Create a new cursor wrapper with metrics.
+    pub const fn new(cursor: C, metrics: Arc<StorageMetrics>) -> Self {
+        Self { cursor, metrics }
+    }
+}
+
+impl<C: OpProofsHashedCursor> OpProofsHashedCursor for HashedCursorWithMetrics<C> {
+    type Value = C::Value;
+
+    fn seek(&mut self, key: B256) -> OpProofsStorageResult<Option<(B256, Self::Value)>> {
+        self.metrics.record_operation(StorageOperation::HashedCursorSeek, || self.cursor.seek(key))
+    }
+
+    fn next(&mut self) -> OpProofsStorageResult<Option<(B256, Self::Value)>> {
+        self.metrics.record_operation(StorageOperation::HashedCursorNext, || self.cursor.next())
+    }
+}
+
+impl<S> OpProofsStorage for OpProofsStorageWithMetrics<S>
+where
+    S: OpProofsStorage,
+{
+    type StorageTrieCursor = TrieCursorWithMetrics<S::StorageTrieCursor>;
+    type AccountTrieCursor = TrieCursorWithMetrics<S::AccountTrieCursor>;
+    type StorageCursor = HashedCursorWithMetrics<S::StorageCursor>;
+    type AccountHashedCursor = HashedCursorWithMetrics<S::AccountHashedCursor>;
+
+    async fn store_account_branches(
+        &self,
+        block_number: u64,
+        updates: Vec<(Nibbles, Option<BranchNodeCompact>)>,
+    ) -> OpProofsStorageResult<()> {
+        let count = updates.len();
+        let start = std::time::Instant::now();
```

Comment:
let's import `Instant` in module prelude. it's useful to see by only reading prelude of module if it has `no-std` support or not.